tracing: add utxocache tracepoints

arnabsen1729 opened the pull request:
This PR adds <count> tracepoints (including documentation).

If you are new to USDTs you are requested to go through [this](https://github.com/bitcoin/bitcoin/pull/22006) to get some context. 

It was proposed in https://github.com/bitcoin/bitcoin/issues/20981#issuecomment-802688809 some useful tracepoints that can be added. On the basis of that, this PR adds the following tracepoints:

<-- create a table -->

### `flush:coins_flush`
Traces every time caches and indexes are flushed.
- `int64` time (in microseconds)
- `uint32` flush mode
- `uint64` coins flushed 
- `uint64` coins memory usage

### `utxo:coins_added`
Traces every time a new coin is added to the cache
- `int32` height
- `int64` value
- `uint32` if the coin is coinbase
- `uint64` coins usage

### `utxo:coins_spent`
Traces every time a coin is spent
- `int32` height
- `int64` value
- `uint32` if the coin is coinbase
- `uint64` coins usage

### `utxo:coins_uncached`
Traces every time a UTXO with the given outpoint is removed from the cache
- `int32` height
- `int64` value
- `uint32` if the coin is coinbase
- `uint64` coins usage

arnabsen1729 committed:
tracing: Tracepoints for coins flush

arnabsen1729 committed:
tracing: tracepoints for utxo set

coins_add, coins_spent, coins_uncached

arnabsen1729 committed:
[tracing] bpftrace script for logging utxo

0xB10C commented on contrib/tracing/log_utxo_coins.py:
```diff
@@ -0,0 +1,116 @@
+  bpf_usdt_readarg(2, ctx, &data->value);
+  bpf_usdt_readarg(3, ctx, &data->isCoinBase);
+  bpf_usdt_readarg(4, ctx, &data->usage);
+  coins_added.perf_submit(ctx, data, sizeof(*data));
```
should be `coins_spent.perf_submit(ctx, data, sizeof(*data));`

arnabsen1729 replied:
Fixed :white_check_mark:

0xB10C commented on contrib/tracing/log_utxo_coins.py:
```diff
@@ -0,0 +1,116 @@
+    bpf_usdt_readarg(2, ctx, &data->value);
+    bpf_usdt_readarg(3, ctx, &data->isCoinBase);
+    bpf_usdt_readarg(4, ctx, &data->usage);
+    coins_added.perf_submit(ctx, data, sizeof(*data));
```
should be `coins_uncached.perf_submit(ctx, data, sizeof(*data));`

arnabsen1729 replied:
Fixed :white_check_mark:

0xB10C commented on doc/tracing.md:
```diff
@@ -129,6 +129,39 @@ Arguments passed:
+It is called *after* a new coin is added to the cache.
+
+Arguments passed:
+1. Height of the block the coin was added to as `int32`
```
I'd find `1. Block height the coin was added to the UTXO-set as 'int32'` clearer.

arnabsen1729 replied:
Updated

0xB10C commented on doc/tracing.md:
```diff
@@ -129,6 +129,39 @@ Arguments passed:
+
+#### Tracepoint `utxo:coins_added`
+
+It is called *after* a new coin is added to the cache.
```
Don't think it's important that the tracepoint is called _after_ the coin was added to the UTXO set cache.

Could be `Is called when a coin is added to the UTXO cache.`

arnabsen1729 replied:
Okay, made the changes

0xB10C commented on doc/tracing.md:
```diff
@@ -129,6 +129,39 @@ Arguments passed:
+
+#### Tracepoint `utxo:coins_spent`
+
+It is called *after* a coin is spent.
```
Same, _after_ is not really relevant here.

0xB10C commented on src/coins.cpp:
```diff
@@ -229,8 +244,17 @@ bool CCoinsViewCache::Flush() {
 void CCoinsViewCache::Uncache(const COutPoint& hash)
 {
     CCoinsMap::iterator it = cacheCoins.find(hash);
+
```
no whitespace changes please :)

0xB10C commented on contrib/tracing/log_utxo_coins.py:
```diff
@@ -0,0 +1,116 @@
+
+""" Example script to log details about add/spend/uncache coin events.
+utilizing USDT probes and the utxo:coins_added, utxo:coins_spent,
+utxo:coins_uncached  tracepoint. """
```
nit: space between `utxo:coins_uncached` and `tracepoint`.

0xB10C commented on contrib/tracing/log_utxo_coins.py:
```diff
@@ -0,0 +1,116 @@
+
+program = """
+# include <uapi/linux/ptrace.h>
+struct coin_data
+{
+  u32 height;
+  long value;
+  u32 isCoinBase;
+  u64 usage;
+};
+
+
+BPF_ARRAY(data_arr, struct coin_data, 1);
```
I don't think you need a `BPF_ARRAY` here. This is only used when we can't fit the struct into the 512 byte eBPF stack. See [doc/tracing.md#ebpf-virtual-machine-limits](https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md#ebpf-virtual-machine-limits).

See e.g. [this](https://github.com/bitcoin/bitcoin/blob/92f3a4b4d0531976c9484f5b5075487e0925b619/contrib/tracing/p2p_monitor.py#L45-L54) from the P2P monitor as an example.

0xB10C commented on src/validation.cpp:
```diff
@@ -2145,6 +2145,7 @@ bool CChainState::FlushStateToDisk(
             if (!CoinsTip().Flush())
                 return AbortNode(state, "Failed to write to coin database");
             nLastFlush = nNow;
+            TRACE4(flush, coins_flushed, nNow.count(), mode, coins_count, coins_mem_usage);
```
Could be multi-line, similar to the other tracepoints, too. And you likely want to cast here too.

0xB10C replied:
Context shouldn't be `flush`. I think `utxocache` is clearer.

arnabsen1729 replied:
Made the refactoring, also added a new parameter of `outpoint.hash`

0xB10C commented on doc/tracing.md:
```diff
@@ -115,6 +115,20 @@ details.
+`CChainState::FlushStateToDisk` called with.
+
+Arguments passed:
+1. Time in micorseconds as `int64`
```
I think our goal should be to log the flush duration and not the timestamp.

0xB10C commented on contrib/tracing/log_coins_flushed.py:
```diff
@@ -0,0 +1,93 @@
+  u64 coins_mem_usage;
+};
+
+BPF_ARRAY(data_arr, struct data_t, 1);
```
same as above. Doesn't need to be an BPF_ARRAY.

0xB10C commented on contrib/tracing/log_utxo_coins.py:
```diff
@@ -0,0 +1,116 @@
+{
+  u32 height;
+  long value;
+  u32 isCoinBase;
```
are booleans supported? (I haven't checked)

0xB10C reviewed (commented):
first round of review

some general notes:
- [x] you can and should use `flake8` and `autopep8` to format Python scripts
- [x] I think we should pass information about the [outpoint](https://github.com/bitcoin/bitcoin/blob/92f3a4b4d0531976c9484f5b5075487e0925b619/src/primitives/transaction.h#L26-L30) (txid and output index) for UTXO set operations too. For passing the txid see arguments 1 and 7 of the `validation:block_connected` tracepoint.
- [x] I think naming the context `utxocache`, `utxosetcache`, or even `coinscache` is clearer. See https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md#adding-tracepoints-to-bitcoin-core
- [ ] ideally, we'd pass fCoinbase as `bool`, however I haven't checked if this is even possible

0xB10C commented:
> I think we should pass information about the outpoint (txid and output index) for UTXO set operations too. For passing the txid see arguments 1 and 7 of the validation:block_connected tracepoint.

An outpoint is a reference to a previous output. It consists of a txid (called `hash` in COutPoint) and the output index of the transaction (called `n` in COutPoint). See [`COutPoint`](https://github.com/bitcoin/bitcoin/blob/92f3a4b4d0531976c9484f5b5075487e0925b619/src/primitives/transaction.h#L25-L30).

I saw you added the `hash` aka txid of the outpoint as a C string. It would be good to pass the txid as bytes too. That could be done in the 7th argument (not accessible by bpftrace, and that is fine as bpftrace currently can't handle bytes arrays properly anyway) similar as we do in the `validation:block_connected` tracepoint. Additionally, we'd need the pass the output index (`n`) too. 

Would be good to log the outpoints in the bcc and bpftrace scripts too. Often outpoints are formatted as `<txid>:<index>`. And they should be documented.

arnabsen1729 committed:
[tracing] added new params to tp and used ctypes

arnabsen1729 committed:
[tracing] index to outpoints added

arnabsen1729 committed:
X[tracing] added txid as bytes to tp param

arnabsen1729 commented:
@0xB10C I have used the `bool` datatype in the BCC script `log_coins_flushed.py` using ctype https://github.com/0xB10C/bitcoin/pull/1/commits/f43febb960beaa3be58269eaccebea14a783a9c4.

Also, updated the tracepoints context and names and added the `txid` as bytes to `add`, `spent` and `uncache` tracepoints.

0xB10C commented on contrib/tracing/log_coins_flushed.py:
```diff
@@ -1,12 +1,13 @@
 #!/usr/bin/env python3
 
 import sys
+import ctypes as ct
```
```suggestion
import ctypes
```

0xB10C commented on contrib/tracing/log_coins_flushed.py:
```diff
@@ -45,12 +49,26 @@
 ]
 
+# define output data structure in Python
+class Data(ct.Structure):
+    _fields_ = [
+        ("ts", ct.c_long),
+        ("mode", ct.c_uint32),
+        ("coins", ct.c_uint64),
+        ("coins_mem_usage", ct.c_uint64),
+        ("is_flush_prune", ct.c_bool),
+        ("is_full_flush", ct.c_bool)
+    ]
+
```
```suggestion
class Data(ctypes.Structure):
    _fields_ = [
        ("ts", ctypes.c_long),
        ("mode", ctypes.c_uint32),
        ("coins", ctypes.c_uint64),
        ("coins_mem_usage", ctypes.c_uint64),
        ("is_flush_prune", ctypes.c_bool),
        ("is_full_flush", ctypes.c_bool)
    ]

```

I'd prefer not to shorten `ctypes` here. As it's 1) not so long and 2) is clearer when reviewing / reading the code.

0xB10C commented on contrib/tracing/log_coins_flushed.py:
```diff
@@ -45,12 +49,26 @@
+# define output data structure in Python
+class Data(ct.Structure):
+    _fields_ = [
+        ("ts", ct.c_long),
```
Could this be a `c_uint64` too?
```suggestion
        ("ts", ct.c_uint64),
```

arnabsen1729 replied:
Yup, there was a type `c_uint64` and it seemed to work without any issue

0xB10C commented on contrib/tracing/log_coins_flushed.py:
```diff
@@ -45,12 +49,26 @@
+    _fields_ = [
+        ("ts", ct.c_long),
+        ("mode", ct.c_uint32),
+        ("coins", ct.c_uint64),
```
```suggestion
        ("coins_count", ct.c_uint64),
```

0xB10C commented on contrib/tracing/log_utxo_coins.py:
```diff
@@ -15,6 +15,7 @@
+  u32 index;
   char hash[MAX_HASH_SIZE];
   long value;
   u32 isCoinBase;
```
should be `bool` too

arnabsen1729 replied:
Did you mean to `isCoinBase`?

Cause, `index` refers to the `COutPoint::n` here https://github.com/0xB10C/bitcoin/blob/7be143a960e2bb9ac81144f55c45731c1bb209c5/src/primitives/transaction.h#L26-L30

For the `isCoinBase` which refers to `Coin::fCoinBase` it is of `unsigned int` type.
https://github.com/0xB10C/bitcoin/blob/7be143a960e2bb9ac81144f55c45731c1bb209c5/src/coins.h#L30-L37

0xB10C replied:
yes, sorry if that was unclear. For me the comment is on the `isCoinbase` line.

![image](https://user-images.githubusercontent.com/19157360/131505950-c04f70b9-bfce-445e-a143-c1b7150fdc81.png)

0xB10C replied:
You could use Coin::IsCoinBase() which returns a bool

0xB10C replied:
See https://github.com/0xB10C/bitcoin/pull/1/files?file-filters%5B%5D=.cpp#r700080149

0xB10C commented on src/validation.cpp:
```diff
@@ -2147,6 +2147,14 @@ bool CChainState::FlushStateToDisk(
             nLastFlush = nNow;
             full_flush_completed = true;
         }
+        TRACE6(utxocache, flush,
+            (GetTimeMicros() - nNow.count()), // in microseconds (µs)
+            mode,
+            coins_count,
+            coins_mem_usage,
+            fFlushForPrune,
+            fDoFullFlush
+        );
```
type castings? :)

0xB10C commented on src/coins.cpp:
```diff
@@ -229,8 +250,20 @@ bool CCoinsViewCache::Flush() {
+            (uint32_t)hash.n,
+            (uint32_t)it->second.coin.nHeight,
+            (int64_t)it->second.coin.out.nValue,
+            (uint)it->second.coin.fCoinBase,
```
```suggestion
            (bool)it->second.coin.fCoinBase,
```

0xB10C replied:
```suggestion
            (bool)it->second.coin.IsCoinBase(),
```

https://github.com/bitcoin/bitcoin/blob/b3a2b8c29fda5d924726d3087535bd34af7012b6/src/coins.h#L55-L57

0xB10C commented on contrib/tracing/log_coins_flushed.py:
```diff
@@ -0,0 +1,109 @@
+
+struct data_t
+{
+  long ts;
```
```suggestion
  u64 duration;
```

0xB10C commented on contrib/tracing/log_coins_flushed.py:
```diff
@@ -0,0 +1,109 @@
+{
+  long ts;
+  u32 mode;
+  u64 coins;
```
```suggestion
  u64 coins_count;
```

0xB10C commented on contrib/tracing/log_coins_flushed.py:
```diff
@@ -0,0 +1,109 @@
+
+int trace_flush(struct pt_regs *ctx) {
+  struct data_t data = {};
+
```
```suggestion
```

0xB10C commented on contrib/tracing/log_utxo_coins.py:
```diff
@@ -0,0 +1,116 @@
+
+    print("Logging Add, Spend and Uncache in the UTXO set. Ctrl-C to end...")
+    print("%-70s %-10s %-10s %-15s %-10s %s" %
+          ("txHash", "Action", "Height", "Satoshis", "Coin Base", "Usage"))
```
```suggestion
          ("Outpoint", "Action", "Height", "Value", "Coinbase", "Memory Usage"))
```

0xB10C replied:
Maybe also add cache size (count of cached utxos)?

0xB10C replied:
In src/coins.cpp you should be able to access cacheCoins in src/coins.h https://github.com/bitcoin/bitcoin/blob/7317e14a44c6efc545e6fb9bcedee7174e93a8fa/src/coins.h#L221

which is a `std::unordered_map` https://github.com/bitcoin/bitcoin/blob/7317e14a44c6efc545e6fb9bcedee7174e93a8fa/src/coins.h#L134

which implements size() https://en.cppreference.com/w/cpp/container/unordered_map/size

0xB10C commented on contrib/tracing/log_utxo_coins.py:
```diff
@@ -0,0 +1,116 @@
+
+
+def print_event(event, action):
+    # print(event.hash.decode('ASCII'), action, event.height, event.value, "YES" if event.isCoinBase else "NO", event.usage)
```
```suggestion
```

0xB10C reviewed (commented):
Good work! I left a few more comments worth addressing. Also, some of the previous comments (e.g. https://github.com/0xB10C/bitcoin/pull/1/files#r692351276) are still open.

0xB10C commented on src/coins.cpp:
```diff
@@ -95,6 +96,15 @@ void CCoinsViewCache::AddCoin(const COutPoint &outpoint, Coin&& coin, bool possi
+        (uint32_t)outpoint.n,
+        (uint32_t)coin.nHeight,
+        (int64_t)coin.out.nValue,
+        (uint)coin.fCoinBase,
```
see https://github.com/0xB10C/bitcoin/pull/1/files?file-filters%5B%5D=.cpp#r700080149

0xB10C commented on src/coins.cpp:
```diff
@@ -120,6 +130,17 @@ bool CCoinsViewCache::SpendCoin(const COutPoint &outpoint, Coin* moveout) {
+        (uint32_t)outpoint.n,
+        (uint32_t)it->second.coin.nHeight,
+        (int64_t)it->second.coin.out.nValue,
+        (uint)it->second.coin.fCoinBase,
```
https://github.com/0xB10C/bitcoin/pull/1/files?file-filters%5B%5D=.cpp#r700080149

0xB10C commented:
closing this in favor of https://github.com/bitcoin/bitcoin/pull/22902 :)
